Update docs to mention CUDA support for Future

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50048 Update docs to mention CUDA support for Future**

To reflect the many changes introduced recently.

In my mind, CUDAFuture should be considered a "private" subclass, which in practice should always be returned as a downcast pointer to an ivalue::Future. Hence, we should document the CUDA behavior in the superclass, even if it's CUDA-agnostic, since that's the interface the users will see also for CUDA-enabled futures.

Differential Revision: [D25757474](https://our.internmc.facebook.com/intern/diff/D25757474/)

lw committed:
Update docs to mention CUDA support for Future

To reflect the many changes introduced recently.

In my mind, CUDAFuture should be considered a "private" subclass, which in practice should always be returned as a downcast pointer to an ivalue::Future. Hence, we should document the CUDA behavior in the superclass, even if it's CUDA-agnostic, since that's the interface the users will see also for CUDA-enabled futures.

Differential Revision: [D25757474](https://our.internmc.facebook.com/intern/diff/D25757474/)

[ghstack-poisoned]

lw commented on docs/source/futures.rst:
```diff
@@ -5,10 +5,6 @@
 torch.futures
 =============
 
-.. warning::
-  The ``torch.futures`` package is experimental and subject to change.
```
I'm proposing to remove it as I suspect this is "stale": the Future class is used by the RPC API, which is now stable, hence I suppose the Future API is stable too?

lw commented on torch/futures/__init__.py:
```diff
@@ -21,35 +21,101 @@ class Future(torch._C.Future, Generic[T], metaclass=_PyFutureMeta):
         """
         return super().wait(non_blocking)
 
+    def value(self) -> T:
```
I'm not adding this method, it already existed (implicitly inherited from the superclass) but it was undocumented.

mrshenli replied:
This LGTM, but I wonder if we should call this `result()` to 1) stay consistent with Python result 2) we already have a `set_result` API.

mrshenli replied:
Yep, just a general question on whether it worths to retire `value` and rename it `result`

lw commented on torch/futures/__init__.py:
```diff
@@ -82,20 +148,19 @@ def then(self, callback):  # type: (Callable[[Future[T]], S]) -> Future[S]
 
     # Have to use string annotations because  PEP-0563 is not available in 3.6
-    def _add_done_callback(self, callback):  # type: (Callable[[Future[T]], None]) -> None
+    def add_done_callback(self, callback):  # type: (Callable[[Future[T]], None]) -> None
```
I am removing the leading underscore from this method because, if it was intentional, I don't think it was really achieving the desired effect. It probably was intended to make this method "private", which it did, but it failed to override the `add_done_callback` method from the superclass. This method was appearing in the docs, however without any documentation (see https://pytorch.org/docs/master/futures.html#torch.futures.Future.add_done_callback). Either we somehow remove/hide the superclass's method, or we keep it but document it. I went for the latter.

I looked for usages of the underscore-prefixed version of this method in the codebase but couldn't find any.

mrshenli replied:
Agree. Leaving `add_done_callback` undocumented wasn't intentional, and we should fix. Both removing this method and making it public look OK to. There was a request from @jjlilley for this feature, and this also complies with Python APIs. There was also an argument for not exposing it, and `then` already covers a superset of use cases.

mrshenli replied:
this is actually necessary in use cases like the one below, right? 

https://discuss.pytorch.org/t/pytorch-distributed-rpc-bottleneck-in-recursive-compile-class/117005/5

pritamdamania87 commented on torch/futures/__init__.py:
```diff
@@ -21,35 +21,101 @@ class Future(torch._C.Future, Generic[T], metaclass=_PyFutureMeta):
         r"""
         Block until the value of this ``Future`` is ready.
 
+        If the value contains tensors that reside on GPUs, then an additional
+        synchronization is performed with the kernels (executing on the device)
+        which may be asynchronously populating those tensors. Such sync can be:
```
I don't think this is true for all futures we return in PyTorch. As per my understanding only ProcessGroupNCCL and our RPC framework guarantee this, but do ops like `torch.jit.fork` respect this? For example, I can do this today with GPU tensors:

```
import torch

@torch.jit.script
def add_one(inp):
    return inp + 1

t = torch.rand(3,3).cuda()
fut = torch.jit.fork(add_one, t)
print(fut.wait())
```

I believe the Future returned by torch.jit.fork will be a CPU future and not a CUDA future despite containing GPU tensors. The non_blocking behavior is controlled by what kind of Future you return and not what kind of tensors the Future holds. Maybe we should update the docs to reflect which APIs would return a CUDA future and the non_blocking flag applies only to those futures.

lw replied:
Not sure if it changed recently, but today it seems that `torch.jit.fork()` just calls `torch._C.fork()`:
https://github.com/pytorch/pytorch/blob/43f6deb6e447aae9d2edbd9441043b03c6ff0992/torch/jit/_async.py#L82

Which in turn is defined as returning a `torch._C.Future`:
https://github.com/pytorch/pytorch/blob/43f6deb6e447aae9d2edbd9441043b03c6ff0992/torch/_C/__init__.pyi.in#L168

Which is (basically) the same class as `torch.futures.Future`.

pritamdamania87 commented on torch/futures/__init__.py:
```diff
@@ -21,35 +21,101 @@ class Future(torch._C.Future, Generic[T], metaclass=_PyFutureMeta):
-        added. The callback must take one argument, which is the reference to
-        this ``Future``. The callback function can use the ``Future.wait()`` API
-        to get the value.
+        the same ``Future``, but the order in which they will be executed cannot
+        be guaranteed (to enforce a certain order consider chaining:
+        ``fut.then(cb1).then(cb2)``). The callback must take one argument, which
```
Not sure I followed this completely, the order **can** be guaranteed as you suggested with chaining. Even if we don't do chaining and do something like `fut.then(cb1);fut.then(cb2)`, the order should be guaranteed right?

mrshenli replied:
IIUC, this order might not be enforced. Suppose we have `fut.then(cb1);fut.then(cb2);fut.then(cb3);`. The fut is marked as completed by a different RPC thread after `fut.then(cb2`. Then, `cb3` will run inline, while the other two cbs will run in the RPC thread. I guess in that case, there is no guarantee on the orders? cc @lw

pritamdamania87 commented on torch/futures/__init__.py:
```diff
@@ -21,35 +21,101 @@ class Future(torch._C.Future, Generic[T], metaclass=_PyFutureMeta):
+        is the reference to this ``Future``. The callback function can use the
+        :meth:`value` method to get the value.
+
+        If the ``Future``'s value contains tensors that reside on GPUs, the
```
As I mentioned above, I don't think this depends on whether the Future contains tensors.

pritamdamania87 commented on torch/futures/__init__.py:
```diff
@@ -121,6 +186,17 @@ def set_result(self, result: T) -> None:
         completed and trigger all attached callbacks. Note that a ``Future``
         cannot be marked completed twice.
 
+        If the result contains tensors that reside on GPUs, this method can be
```
Same as above.

lw committed:
Update on "Update docs to mention CUDA support for Future"


To reflect the many changes introduced recently.

In my mind, CUDAFuture should be considered a "private" subclass, which in practice should always be returned as a downcast pointer to an ivalue::Future. Hence, we should document the CUDA behavior in the superclass, even if it's CUDA-agnostic, since that's the interface the users will see also for CUDA-enabled futures.

Differential Revision: [D25757474](https://our.internmc.facebook.com/intern/diff/D25757474/)

[ghstack-poisoned]

lw committed:
Update on "Update docs to mention CUDA support for Future"


To reflect the many changes introduced recently.

In my mind, CUDAFuture should be considered a "private" subclass, which in practice should always be returned as a downcast pointer to an ivalue::Future. Hence, we should document the CUDA behavior in the superclass, even if it's CUDA-agnostic, since that's the interface the users will see also for CUDA-enabled futures.

Differential Revision: [D25757474](https://our.internmc.facebook.com/intern/diff/D25757474/)

[ghstack-poisoned]

lw committed:
Update on "Update docs to mention CUDA support for Future"


To reflect the many changes introduced recently.

In my mind, CUDAFuture should be considered a "private" subclass, which in practice should always be returned as a downcast pointer to an ivalue::Future. Hence, we should document the CUDA behavior in the superclass, even if it's CUDA-agnostic, since that's the interface the users will see also for CUDA-enabled futures.

Differential Revision: [D25757474](https://our.internmc.facebook.com/intern/diff/D25757474/)

[ghstack-poisoned]

mrshenli commented on torch/futures/__init__.py:
```diff
@@ -45,30 +47,86 @@ def done(self) -> bool:
+        reside on a GPU, it can do so even if the kernels that are producing
+        these tensors are still running on the device, as long as the callback
+        didn't change streams during its execution. If one wants to change
+        streams, one must be careful to re-synchronize them with the original
```
it might worth provide an example of how this can be done. This can be added in follow up PRs, or in Future/CUDA-RPC best practices.

mrshenli reviewed (approved):
LGTM

mrshenli reviewed (commented):
mypy failure is real though:

```
torch/futures/__init__.py:99: error: "value" undefined in superclass  [misc]
```

lw commented on torch/futures/__init__.py:
```diff
@@ -21,35 +21,101 @@ class Future(torch._C.Future, Generic[T], metaclass=_PyFutureMeta):
+
+        Args:
+            non_blocking (bool): whether the sync with the GPU kernels should be
+                                 non-blocking.
```
The `non_blocking` flag has been removed.

lw commented on torch/futures/__init__.py:
```diff
@@ -21,35 +21,101 @@ class Future(torch._C.Future, Generic[T], metaclass=_PyFutureMeta):
+        ``value()`` could fail.
+
+        If the value contains tensors that reside on GPUs, then this method will
+        *not* perform any additional synchronization. This should be done
```
Italics, according to the docs: https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html

lw committed:
Update on "Update docs to mention CUDA support for Future"


To reflect the many changes introduced recently.

In my mind, CUDAFuture should be considered a "private" subclass, which in practice should always be returned as a downcast pointer to an ivalue::Future. Hence, we should document the CUDA behavior in the superclass, even if it's CUDA-agnostic, since that's the interface the users will see also for CUDA-enabled futures.

Differential Revision: [D25757474](https://our.internmc.facebook.com/intern/diff/D25757474/)

[ghstack-poisoned]
